core/src/main/scala/zio/jdbc/ZConnection.scala

```diff
@@ -28,67 +28,35 @@ import java.sql.{ Blob, Connection, PreparedStatement, Statement }
-              case v: Int                   => statement.setInt(paramIndex, v)
-              case v: Long                  => statement.setLong(paramIndex, v)
-              case v: Short                 => statement.setShort(paramIndex, v)
-              case v: Byte                  => statement.setByte(paramIndex, v)
-              case v: Char                  => statement.setString(paramIndex, v.toString)
-              case v: Double                => statement.setDouble(paramIndex, v)
-              case v: Blob                  => statement.setBlob(paramIndex, v)
-              case v: java.math.BigDecimal  => statement.setBigDecimal(paramIndex, v)
-              case v: scala.math.BigDecimal => statement.setBigDecimal(paramIndex, v.bigDecimal)
-              case v: java.math.BigInteger  => statement.setBigDecimal(paramIndex, new java.math.BigDecimal(v))
-              case v: scala.math.BigInt     => statement.setBigDecimal(paramIndex, new java.math.BigDecimal(v.bigInteger))
-              case v: java.sql.Date         => statement.setDate(paramIndex, v)
-              case v: java.sql.Time         => statement.setTime(paramIndex, v)
-              case v: java.sql.Timestamp    => statement.setTimestamp(paramIndex, v)
-              case v: Boolean               => statement.setBoolean(paramIndex, v)
-              case v: Float                 => statement.setFloat(paramIndex, v)
-              case v                        => statement.setObject(paramIndex, v)
-            }
-
-            paramIndex += 1
-
-          case Syntax(_) =>
-        }
-        i += 1
-      }
-
-      f(statement)
-    }.tapErrorCause { cause =>
-      ZIO.logAnnotate("SQL", sql.toString)(ZIO.logError(s"Error executing SQL due to: ${cause.prettyPrint}"))
+  private[jdbc] def executeSqlWith[A](
+    sql: Sql[_]
+  )(f: PreparedStatement => ZIO[Scope, Throwable, A]): ZIO[Scope, Throwable, A] =
+    accessZIO { connection =>
+      for {
+        statement <- ZIO.acquireRelease(ZIO.attempt {
+                       val sb = new StringBuilder()
+                       sql.foreachSegment(syntax => sb.append(syntax.value))(_ => sb.append("?"))
+                       connection.prepareStatement(sb.toString, Statement.RETURN_GENERATED_KEYS)
+                     })(statement => ZIO.attemptBlocking(statement.close()).ignoreLogged)
+        _         <- ZIO.attempt {
+                       var paramIndex = 1
+                       sql.foreachSegment(_ => ()) { param =>
+                         param.setter(statement, paramIndex, param.value)
+                         paramIndex += 1
+
+                       }
+                     }
+        result    <- f(statement)
+      } yield result
+    }.tapErrorCause { cause => // TODO: Question: do we want logging here, switch to debug for now
+      ZIO.logAnnotate("SQL", sql.toString)(
+        ZIO.logDebugCause(s"Error executing SQL due to: ${cause.prettyPrint}", cause)
```

Comment:
Changing the logging level to `Debug`, but I wonder if there is a way to expose SQL in errors returned by the library?